### ISSUE TITLE
Upload large file

### ISSUE THREAD

**Issue by palashcn** [Ziggeo/ZiggeoPhpSdk#27]
Hi,

When I tried 2.5 gb video file (32057-usbje4129.mov) 
I got this error:

Uncaught ZiggeoException: Gateway Timeout (504) in /var/www/html/ziggeo/classes/ZiggeoConnect.php:43
Stack trace:
#0 /var/www/html/ziggeo/classes/ZiggeoConnect.php(72): ZiggeoConnect->assert_state(Array, 504, false)
#1 /var/www/html/ziggeo/classes/ZiggeoConnect.php(77): ZiggeoConnect->post('/videos/', Array, Array)
#2 /var/www/html/ziggeo/classes/ZiggeoVideos.php(66): ZiggeoConnect->postJSON('/videos/', Array)
#3 /var/www/html/test_video.php(16): ZiggeoVideos->create(Array)
#4 {main}
========
is there any video file size limit or video type?

Thanks

**Comment by Bane-D** [comment]
Hi,

504 Gateway Timeout error as with 5xx errors in general can be a bit misleading suggesting that it is end server error, however there can be few other places the issue can actually happen, so there are few points to be aware of.

In regards to 504 error, can you please make sure to check if:

1. You can see if your webserver error log (like Apache error log) contains more details
2. Same with the PHP error log
3. If you are making calls through any sort of proxy
  * Try going directly to see what happens (making sure that proxy is not causing the error)
4. Please make sure that it is not a DNS Issue
  * You can do this by trying to send few smaller files. This would allow you to see that you can send videos and reach our system in general

Our system is used all the time and we do not impose any limits in terms of time it takes or size it is of, however different parts of your system can. For example PHP has a small timeframe to execute scripts after which it is just stopped through a timeout error. This is by default and requires you to increase it. Similar is the case with the file size.

Apache or any other webserver that you use, like nginx can be set to be used as proxy or gateways or used together with another webserver and if they have a time limit it could be reached.

So long story short, we do not impose any limits unless you set them up yourself. I can also not see any error on our side. Please do try the steps above just to see if we can see more details on what is happening.

**Comment by palashcn** [comment]
Hi,

I have tried with small file and those files are uploaded easily. so I have
increased PHP max_execution_time, max upload file size, post_max_size,
memory_limit, socket_execution_time, apache timeout but did not get any luck
As I'm using Ubuntu 18.04  PHP 7.2 server.

On the other side, I have tried the same file upload through PHP curl and
it's working well here, so I am confident that this is an issue in Ziggeo
server API end w.

here is my code which worked successfully in our server with 2.3 gb file

<?php
$destination = "/home/ubuntu/32057-usbje4129.mov";

$curl = curl_init("http://mysite.com/test_video2.php");
curl_setopt($curl, CURLOPT_FAILONERROR, true);

if( (ini_get('safe_mode') === 'Off' || ini_get('safe_mode') === false) &&
ini_get('open_basedir') === '' ) {
    //we can only make curl follow the redirects if the safe mod is off and
open_basedir is not set, otherwise it would throw error
    curl_setopt($curl, CURLOPT_FOLLOWLOCATION, true);
}
curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, true);
curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, 60000);
curl_setopt($curl, CURLOPT_TIMEOUT, 60000);
curl_setopt($curl, CURLOPT_POST, true);
$data["file"] = new CurlFile(str_replace("@", "", $destination),
"video/mp4", "video.mp4");
curl_setopt($curl, CURLOPT_POSTFIELDS, $data);

$result = curl_exec($curl);
if($result === false)
{
    echo 'Curl error: ' . curl_error($curl);
}
curl_close($curl);

Thanks & Regards,

Palash Pramanick
Tech Lead
Capital Numbers
EN 66, 3rd Floor, Salt Lake City, Sector V
Kolkata 700091 West Bengal, India

ISO 9001:2008 | D&B D-U-N-S Number: 65-080-6800 | Google Partner
<https://www.google.com/partners/?hl=en-US#a_profile;idtf=7716899300;>
Why Hire Remote Staff from Capital Numbers?
<https://www.youtube.com/watch?v=w-zu1uO4pHg> (YouTube)
Social : LinkedIn
<https://www.linkedin.com/company/capitalnumbers-infotech-pvt-ltd-> |
Twitter <https://twitter.com/_CNInfotech> | Facebook
<https://www.facebook.com/CapitalNumbers> | Google+
<http://plus.google.com/+CapitalNumbers> | SlideShare
<http://www.slideshare.net/capitalnumbers> | Infographics
<http://visual.ly/users/capitalnumbers>


Skype: palash.cn
Gtalk/Gmail: palash@capitalnumbers.com



This email and any files transmitted with it are confidential and intended
solely for the use of the individual or entity to whom they are addressed.
If you have received this email in error please notify the system manager.
This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system. If you are not the intended recipient
you are notified that disclosing, copying, distributing or taking any
action in reliance on the contents of this information is strictly
prohibited.

**Comment by Bane-D** [comment]
Hi Palash,

Thank you for the additional info.

We are actually also using curl in the SDK, and the code above you posted is taken from our SDK [/classes/ZiggeoConnect.php](/Ziggeo/ZiggeoPhpSdk/tree/master/classes/ZiggeoConnect.php)

The only difference is that you are adding the file in the following manner:
> $data["file"] = new CurlFile(str_replace("@", "", $destination), "video/mp4", "video.mp4");

With that said, could you please let me know the code you are using originally with our own SDK functions/calls just to see if anything seems off there, since it should otherwise work. 

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane-D,

Yes, I have used that class and those codes are taken from your SDK.

And my server is using the apache.

So let me know.

Thanks
Paalsh

**Comment by Bane-D** [comment]
Hi Palash,

I just finished with a test upload (larger file) and it went through fine on my end. I am using Kubuntu myself however that is fairly similar to what you have.

Can you please go to the demo folder of our SDK and copy the video file to the same to make it easier, then execute the following within your console:

`php videos_create.php --app_token "1234" --private_key "567890" --file_path "./32057-usbje4129.mov"`

Of course replace _1234_ with your actual application token and _567890_ with your actual private key.

Do let me know how it goes.

**Comment by palashcn** [comment]
Hi Bane,

I have tried with that command and found this issue:
PHP Fatal error:  Uncaught ZiggeoException: Continue (100) in /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php:121
Stack trace:
#0 /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php(146): ZiggeoConnect->assert_state(Array, 100, Array)
#1 /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php(151): ZiggeoConnect->post('/v1/videos/', Array, Array)
#2 /var/www/html/ziggeo_latest/classes/ZiggeoVideos.php(70): ZiggeoConnect->postJSON('/v1/videos/', Array)
#3 /var/www/html/ziggeo_latest/demos/videos_create.php(17): ZiggeoVideos->create(Array)
#4 {main}
  thrown in /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php on line 121

I tried in the localhost and my production AWS server and both have same issue.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

OK thank you. So based on that everything is fine at that point. To have someone else explain the code please see here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/100

Basically that means that the PHP should continue sending the file. In reality unless you have some setup that requires this to stop the execution, PHP should just continue with the upload. This is basically our server saying all is good, continue sending file if you have anything else to send our way. At that point PHP should reply with the file data or the header that file upload is finalized.

* I did not look to be precise here, just to explain the general process.

Not sure why our code is even getting info about the 100 code being replied.

Now with that said, let's try something. I have made and pushed the code that allows you to follow the upload progress and see what is happening.

You will see the info like this:
> Runtime: 41.42319393158 seconds Upload status: 92.636700630188 / 892.636700630188 mb

At the very end, you might see the same size repeated, which is fine, just wait for the console to return, after all with large files there are protocols that might require some additional bytes to be sent.

Now this should give us some info on when the issue happens for you.

It is currently enabled in the config file, however if available later it will be disabled. To use it please download the following code: https://github.com/Ziggeo/ZiggeoPhpSdk/tree/php_progress

Once you do please run the same code as earlier:

`php videos_create.php --app_token "1234" --private_key "567890" --file_path "./32057-usbje4129.mov"`

Please let me know how it goes.

**Comment by palashcn** [comment]
Hi Bane,

it's showing the same issue, here is output
**when "request_timeout" => 60**
Runtime: 59.995982885361 seconds Upload status: 314.609375 / 2127.0294265747 mb
Runtime: 59.996011972427 seconds Upload status: 314.609375 / 2127.0294265747 mb
Runtime: 59.996224880219 seconds Upload status: 314.625 / 2127.0294265747 mb
Runtime: 59.996244907379 seconds Upload status: 314.625 / 2127.0294265747 mb
Runtime: 59.999418020248 seconds Upload status: 314.625 / 2127.0294265747 mbPHP Fatal error:  Uncaught ZiggeoException: Continue (100) in /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php:155
Stack trace:
#0 /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php(180): ZiggeoConnect->assert_state(Array, 100, Array)
#1 /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php(185): ZiggeoConnect->post('/v1/videos/', Array, Array)
#2 /var/www/html/ziggeo_latest/classes/ZiggeoVideos.php(70): ZiggeoConnect->postJSON('/v1/videos/', Array)
#3 /var/www/html/ziggeo_latest/demos/videos_create.php(17): ZiggeoVideos->create(Array)
#4 {main}
  thrown in /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php on line 155

===================
**when "request_timeout" => 600000**
Runtime: 235.75380301476 seconds Upload status: 2127.0294265747 / 2127.0294265747 mb
Runtime: 235.75384306908 seconds Upload status: 2127.0294265747 / 2127.0294265747 mb
Runtime: 235.84935808182 seconds Upload status: 2127.0294265747 / 2127.0294265747 mbPHP Fatal error:  Uncaught ZiggeoException: Gateway Timeout (504) in /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php:155
Stack trace:
#0 /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php(180): ZiggeoConnect->assert_state(Array, 504, Array)
#1 /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php(185): ZiggeoConnect->post('/v1/videos/', Array, Array)
#2 /var/www/html/ziggeo_latest/classes/ZiggeoVideos.php(70): ZiggeoConnect->postJSON('/v1/videos/', Array)
#3 /var/www/html/ziggeo_latest/demos/videos_create.php(17): ZiggeoVideos->create(Array)
#4 {main}
  thrown in /var/www/html/ziggeo_latest/classes/ZiggeoConnect.php on line 155

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Thank you for the details and for trying it out.

I am glad to see that the 100 one is passed by increasing the timeout time which makes sense (as all is good). Now the 504 is still weird, so would be good if you can tell me if you are going through any proxy server?

I have also asked my colleague to check the raw logs on our side related to your account (this can take a bit) to see what might be happening. This should allow us to notice if there is anything somewhere from our side causing it or if not.

Please let me know about proxy and I will keep you updated on what we find out.

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

I did not use any proxy, I tried in my other colleague's PC  to run the script and found the script returning the same error.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Thank you for the wait. My colleague could not find the reference to error on our end. However looking into this we do have an idea on what is actually happening.

Seems that your PHP is actually blocking the common upload through PHP. To confirm this I would like to ask you to try adding a very simple change and try uploading again. If this works for you then I will update our SDK so that it is available in a better way (detecting this and using different ways for uploads).

The change is very simple, just add the following to your ZiggeoConnect.php file:
`curl_setopt($curl, CURLOPT_SAFE_UPLOAD, false);`

Please add it to line 40: https://github.com/Ziggeo/ZiggeoPhpSdk/blob/master/classes/ZiggeoConnect.php#L40

This is how it would look that area:
```
...
        curl_setopt($curl, CURLOPT_USERPWD, $this->application->token() . ":" . $this->application->private_key());
        curl_setopt($curl, CURLOPT_SAFE_UPLOAD, false);
        
        //Check if we should activate progress info output
        if($this->config->get('info')['progress_show'] === true) {
...
```

Hope this solves it for you and looking forward to what you say :)

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

The same error appears, but here I got interesting facts, the uploaded script in loop i.e. when I run the script it's 4 times run and starting from 0 MB again and again and after some time timeout error appeared.

Please try with 2.5 GB video file.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Thank you for trying and helping us get more details on this. I have tried and saw 100 error, and have the code to handle that. With that said I actually noticed something and looks like the same thing you have.

Our SDK is set to re-upload the file on fail. I was able to replicate that by setting the execution time to small number (smaller than it would take me to actually upload). At that time it would just re-do it few times until it ends up failing.
* This is expected. We want you to be in control of execution time and also to make sure that if it fails for some network reason for example for it to try re-uploading again (how many times also depends on your config).

Now since you are uploading very large files the solution could actually be as simple as changing this line:
https://github.com/Ziggeo/ZiggeoPhpSdk/blob/master/classes/ZiggeoConfig.php#L14
Please make that 0 so it looks like so:
`"request_timeout" => 0`

Once you do try again and Curl will not be set to stop the connection after 60 seconds that is default.

Now if you prefer to not have 0 there, you could place some other amount of time to be your default. Please do keep in mind your file sizes and your network speed and adjust it to also have some buffer. So setting time to be higher than needed, would always be a better choice.
This is also the intended purpose of the config file and this setting :)

Please let me know how it goes :)

Bane

**Comment by palashcn** [comment]
Hi Bane,

Now the video uploaded but it took 180 seconds for update and after that video token returned at 403 seconds.

for 4 GB file, it took approx 415 seconds but the video token returned at 816 seconds.

Is that ok?

Thanks
Palash

**Comment by palashcn** [comment]
HI Bane,

But there is now a strange issue, all 3 of my videos showing failed in my dashboard.

54f85c2803db744731ddef9954b480ae
b957b98018cf59b3ee131f831cf221ba
30c836a8eaa62147c383a95609679897

those are 3 videos which I just added recently and showing it's status failed.

Please help.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Looking into this. Both me and my colleague have been uploading files of different sizes today trying to recreate the issue.

With that said, I am glad that it is uploading for you now.

As soon as I get a bit more details about the videos, following up with you.

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

Did you get the issue of why it's showing status failed?

I'm waiting for your reply.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

While we are looking into what is happening exactly and why allowing us to stop it from happening in future one of my colleagues will be adding a new way of uploading files for our system.

This should result in faster uploads and would not be limited in any way to the sizes of the files.

Will update you as soon as I have the update from my colleague on this.

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

Now again, videos are not uploading, could you please give us a permanent solution.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Thank you for letting us know that it stopped working for you at that point again. As I mentioned we are working on the new upload method that will work much better for the large files. The small files are not really affected by this just the giant ones.

So for now, please do use 0 as I mentioned earlier. This is the best way to make sure that the timeout does not happen on the CURL side of things.

Also for now please expect that it might fail with large uploads. Do note of course that we are working on it to make the new approach available as soon as it is possible so that it does not fail any more.

Will keep you updated as soon as this is released. I am not yet sure if this will be released in a way that will require SDK update or not, however will update you as soon as it is available and if needed update the SDK to use this new approach :)

I know it is not equal to "all fixed" however I do hope it helps

Regards,
Bane

**Comment by Bane-D** [comment]
Hi Palash,

Can you please try a new branch to see how it works for you? It works awesome in our tests, however would be happy to hear the same from your side as well.

This is the branch "[signed-upload-test](https://github.com/Ziggeo/ZiggeoPhpSdk/tree/signed-upload-test)".

You would make the same call as you did so far, it will just use a different approach internally.

Looking forward to hear back.

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

Now I'm getting this error with latest code

Runtime: 0.34500002861023 seconds Upload status: 0.00015544891357422 / 0.00015544891357422 mbPHP Fatal error:  Uncaught ZiggeoException: Not Found (404) in /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php:202
Stack trace:
#0 /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php(227): ZiggeoConnect->assert_state(Array, 404, false)
#1 /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php(232): ZiggeoConnect->post('/v1/videos//eff...', Array, Array)
#2 /var/www/html/ziggeo_latest2/classes/ZiggeoVideos.php(48): ZiggeoConnect->postJSON('/v1/videos//eff...', Array)
#3 /var/www/html/test_video.php(19): ZiggeoVideos->apply_effect('', Array)
#4 {main}
  thrown in /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php on line 202

Please let me know the cause.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Sorry, should have mentioned that can happen.

Not all of our servers are updated with the code during the test phase. So if you reach out and it is the server that does not have the code that works with this code it will fire a 404.

Usually we push updates to all of our servers over weekend. So please try again and immediately again if 404 happens. If it still does not seem to work for you please let me know. I will update you once all of the servers support this.

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

I just tried again and getting same issue.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Just to be safe, can you please let me know the code you use to test with? I would like to test with same code and it looks like you might not be using the videos.create for your test, rather the apply effects method.

Regards,
Bane

**Comment by palashcn** [comment]
require_once('ziggeo_latest2/Ziggeo.php');
$ziggeo = new Ziggeo('key', 'key2', 'kry3');

$destination = "/home/ubuntu/2-5-gb.mov"; 
$video = $ziggeo->videos()->create($arguments = array('file' => $destination));

$videoToken = (isset($video->token) ? $video->token : '');

$result = $ziggeo->videos()->apply_effect($videoToken, array(
	"effectprofiletoken" => "_watermark"
));

var_dump($videoToken);
var_dump($result);

**Comment by Bane-D** [comment]
Ah OK, thank you Palash I understand now. So it is videos create however then you are trying to also apply the effect on it. However since the video is actually a 404 error it will of course not be able to create the effect profile.

Will see when all of our servers are updated so that you can check without the 404 being thrown and update you here.

Thank you again for details and checking.

PS: You could change part of the code to avoid further processing on error with something like:
```
if(isset($video->token)) {
    $videoToken = $video->token;

    $result = $ziggeo->videos()->apply_effect($videoToken, array( "effectprofiletoken" => "_watermark"));
}
```

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

I have just tried to upload a video with same code but still getting the same error, is still migration pending?

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

I just checked and the code is actually pushed to all of our servers at this time, both US and EU regions.

So can you please try again and let me know what error you get?

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

I'm getting this error

PHP Fatal error:  Uncaught ZiggeoException: Not Found (404) in /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php:202
Stack trace:
#0 /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php(227): ZiggeoConnect->assert_state(Array, 404, false)
#1 /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php(232): ZiggeoConnect->post('/v1/videos//eff...', Array, Array)
#2 /var/www/html/ziggeo_latest2/classes/ZiggeoVideos.php(48): ZiggeoConnect->postJSON('/v1/videos//eff...', Array)
#3 /var/www/html/migration/rackspace-to-ziggeo.php(148): ZiggeoVideos->apply_effect('', Array)
#4 {main}
  thrown in /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php on line 202


Thanks
Palash

**Comment by Bane-D** [comment]
Thank you Palash,

Can you please try just uploading the video - without the effect profiles being applied and then see how that goes for you?

Looking at the error it looks like the error comes from `ZiggeoVideos->apply_effect('', Array)` which seems to be an empty string that is passed instead of token.

Based on the codes you sent earlier, could you please run the following code example:
```php
<?php
require_once('ziggeo_latest2/Ziggeo.php');
//Please change the key references with actual keys
$ziggeo = new Ziggeo('key', 'key2', 'kry3');

//Please update the file path if needed
$destination = "/home/ubuntu/2-5-gb.mov";
$video = $ziggeo->videos()->create(array('file' => $destination));
var_dump($video);
?>
```

If all goes right you should see the video data being printed in the console.

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

Yes, without effect it's working fine, but we need that effect.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

Awesome :) Yes of course, just trying to locate where the issue was :)

Now, can you please try the following code:

```php
<?php
require_once('ziggeo_latest2/Ziggeo.php');
//Please change the key references with actual keys
$ziggeo = new Ziggeo('key', 'key2', 'kry3');

//Please update the file path if needed
$destination = "/home/ubuntu/2-5-gb.mov";
$video = $ziggeo->videos()->create(array('file' => $destination));

if(isset($video->token)) {
    $result = $ziggeo->videos()->apply_effect(
        $video->token,
        array("effectprofiletoken" => "_watermark")
    );
}
?>
```

Do let me know how it goes.

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

Video token is blank

Please let me know why it's showing blank?

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

One of the reasons you should never copy paste any part of the code unless it is something you have tested out first even if it looks right :)

Please try this:
```php
<?php
require_once('ziggeo_latest2/Ziggeo.php');
//Please change the key references with actual keys
$ziggeo = new Ziggeo('key', 'key2', 'kry3');

//Please update the file path if needed
$destination = "/home/ubuntu/2-5-gb.mov";
$video = $ziggeo->videos()->create(array('file' => $destination));

if(isset($video['token'])) {
    $result = $ziggeo->videos()->apply_effect(
        $video['token'],
        array("effectprofiletoken" => "_watermark")
    );
    var_dump($result);
}
?>
```
The reason why the other looked right yet was not working is because we were trying to use $video->token, which would work if the $video is object, however since we are dealing with the associated arrays, we actually need to use Array approach instead.

Do of course let me know how it goes :)

Bane

**Comment by palashcn** [comment]
Hi Bane, 

So should I used array in my all code for token?

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

I would say yes, however do test the code above to see if that works for you. If it does (and it should), then just go through your code and make the changes where needed.

Technically you could turn that data into class and get the details as object instead, so depending on what you do with the data that can be both true and false.

In general if you try something like:
```php
var_dump($video->token); //Should tell you some object related error or be null (by default)
var_dump($video['token']); //Should return you the string as token
```
You should get an error for one and not for the other. By default this should be associated array and you can always check with the standard PHP functions like [is_array()](https://www.php.net/manual/en/function.is-array.php) or [is_object()](https://www.php.net/manual/en/function.is-object.php) and then go from there :)

Regards,
Bane

**Comment by palashcn** [comment]
Hi Bane,

So this code will be solved permanently or it's just temporary?
Please let me know.

Thanks
Palash

**Comment by Bane-D** [comment]
Hi Palash,

The codes are available on all of our servers however only the new branch has the codes to talk with servers over these new nodes. We will be making the change permanent however want to update all of our server side SDKs.

The master branch should be updated soon, will let you know as soon as it is updated :)

Regards,
Bane

**Comment by palash14** [comment]
Hi Bane,

getting this issue now whenever we trying upload video.
PHP Fatal error:  Uncaught ZiggeoException: Forbidden (403) in /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php:66
Stack trace:
#0 /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php(251): ZiggeoConnect->curlUploadFile('https://ziggeoa...', '/var/www/html/v...', Array)
#1 /var/www/html/ziggeo_latest2/classes/ZiggeoVideos.php(69): ZiggeoConnect->postUploadJSON('/v1/videos-uplo...', 'video', Array, 'video_type')
#2 /var/www/html/test_video.php(18): ZiggeoVideos->create(Array)
#3 {main}
  thrown in /var/www/html/ziggeo_latest2/classes/ZiggeoConnect.php on line 66

Could you please tell me what the reason.

Thanks
Palash
